### ISSUE TITLE
Make T3 throw or warn when attempting to retrieve a service that doesn't exist in debug mode

### ISSUE THREAD

**Issue by jdivock** [box/t3js#89]
Recently we discovered a ~6mo old bug in our system that was able to persist because T3 silently allowed a features flip service that didn't exist to be retrieved and used as a falsey value in subsequent logic. This could have been identified if T3 had thrown upon attempting to fetch the service, or at least dumped to conosle.warn when in debug mode to at least let a developer know they're attempting to access a service that does not exist.

For code cleanliness console statements can then later be stripped out during our build process with the drop_console flag when the code is minified for production use.

**Comment by nzakas** [comment]
As an example, this occurred because of a format similar to this:

```js
var foo = application.getService('foo'),
    bar = foo && foo.get('bar');

if (bar) {
    // do something
}
```

So `foo` was `null` therefore `bar` also became `null`. The problem is that someone was using `bar` in a position where falsiness made no sense.

The decision to have `getService()` return `null` for unknown services was a design decision to allow condition loading of services such that code could test to see if a service was available before attempting to use it.

The question here is what, if anything, should be done to help protect against this occurrence.

**Comment by j3tan** [comment]
Maybe we should introduce something like hasService() to take care of conditional service includes?

**Comment by nzakas** [comment]
That would mean each call to `getService()` would necessarily need to be preceded by `hasService()` in order to avoid an error. That's quite a bit of overhead, especially if people are relying on this behavior right now. So the question there is if this is important enough to inflict pain on existing users or not?

As a side note, we do throw errors if a module or behavior is missing, so we have some precedent there.

**Comment by j3tan** [comment]
I was thinking getService() always throws an error, but if you conditionally want to use a service, you use hasService() first. Thus, most of the times you would use getService() without hasService() since your code will break if a service you depend on doesn't exist.

**Comment by j3tan** [comment]
Arguably, this may lead to something like 
var foo = application.hasService('foo') && application.getService('foo'),
    bar = foo && foo.get('bar');

We rarely conditionally include services so it might be better to treat it like the edge case (with hasService as an out)

**Comment by nzakas** [comment]
Yeah, that's what I'm referring to. Note that this pattern wouldn't have actually prevented the original error (since `foo` is false, `bar` is also false). This is why I'm having a hard time rationalizing to myself that throwing an error is the right solution to this problem. 

Also, keep in mind that we only throw errors in debug mode, so in production mode, `getService()` would still return `null` (or more likely `undefined`).

**Comment by j3tan** [comment]
Perhaps the sample code above was poorly written, lets say we eliminate the whole `&&` assignment issue:

var bar;
// Code now indicates it expects the possibility of a missing service
if (application.hasService('foo')) {
    var foo = application.getService('foo');
    bar = foo.get('bar');
}

On production, we just swallow errors like missing modules/behaviors and fire a message instead. Throwing an error on missing services would do the same thing (but now we could log them). The variable would then be assigned to `undefined` on production, as you indicated. It then boils down to the current framework expecting developers to handle every getService() correctly when it doesn't return anything. 

Throwing a specific error here will let us catch issues faster on dev and it would eliminate the `foo && foo.get()` pattern since we would expect the service to exist. Obviously, this would be a breaking change that would require us to refactor several code paths, but overall, it might be a safer approach going forward.

You are correct in that throwing an error wouldn't have caught the issue in our current code base, but I think that is a side effect of promoting the `foo && foo.get()` pattern which would erroneously return a falsy value. If we promoted the more explicit, optional service pattern, `foo.get()` simply would have failed since the service object wouldn't exist.

**Comment by nzakas** [comment]
If the problem is purely one of promoted patterns, then we could also disallow the pattern in our own code base (via an ESLint rule).

**Comment by reyuto** [comment]
Why not have the getService throw an error in the dev mode, and then for cases which need to conditionally add a service, they can use a try catch block to do so. This way they can handle the error in their code, and except, there won't be any error in production and it would not work as designed. I know this is not a solution, but still putting it up here, in case it sparks another idea.

**Comment by davidtongbox** [comment]
I often do this in my code:
```javascript
var myService = application.getService('my-service');

if (myService) {
...
}
```
I almost always do this whenever I need a service that is not "core" (a "ajax" service is almost definitely present, while a "snack-bar" service might not).

Using this pattern, I don't need a hasService() method. It also encourages me to keep a reference to a service. However, I recognize that this won't address the issue discussed in this topic because a pattern doesn't do much to alert developers about a missing service.

I do prefer this to throwing an error though because writing a custom try-catch wherever we get a service is not just repetitive but the catch code is hard to be standardized.

**Comment by mattbasta** [comment]
We should always throw an error. This is the behavior that virtually every AMD-like system out there uses, and having different behavior is yet another difference from what new users would expect. Not throwing an exception in production means that problems related to asset compilation would fail silently or throw different and unrelated exceptions.

**Comment by jdivock** [comment]
I'll just reiterate my support for throwing when a service does not exist. It's still unclear to me why T3 has chosen to roll it's own dependency injection system, but the current convention does not align with any of the other patterns out there as pointed out by @mattbasta (angular DI, commonJS, AMD, etc). Fetching a service that may not exist should be seen as a dangerous operation and should be handled in a try/catch.

**Comment by j3tan** [comment]
Judging from the feedback and expected developer behavior, we will be changing getService() to throw an error. This is a breaking change and will be merged in with 2.0.0 along with some other breaking changes.

In addition, I'll be adding the hasService() method to avoid writing try/catch blocks. You typically want to save try/catch for actual errors, not 'optional' expected behavior.
